pkg/controller/controller.go

```diff
@@ -0,0 +1,285 @@
+}
+
+type remoteCluster struct {
+	stopCh              chan struct{}
+	clusterID           string
+	serviceInformer     cache.SharedIndexInformer
+	queue               workqueue.RateLimitingInterface
+	remoteServices      remoteServiceMap
+	federator           federate.Federator
+	remoteServicesMutex *sync.Mutex
+}
+
+func New(federator federate.Federator) *LightHouseController {
+	return &LightHouseController{
+		federator:           federator,
+		remoteServices:      make(remoteServiceMap),
+		remoteClusters:      make(remoteClustersMap),
+		remoteClustersMutex: &sync.Mutex{},
+
+		newClientset: func(c *rest.Config) (kubernetes.Interface, error) {
+			return kubernetes.NewForConfig(c)
+		},
+	}
+}
+
+func (r *LightHouseController) Start() error {
+	err := r.federator.WatchClusters(r)
+	if err != nil {
+		return fmt.Errorf("could not register cluster watch: %v", err)
+	}
+
+	return nil
+}
+
+func (r *LightHouseController) Stop() {
+	r.remoteClustersMutex.Lock()
+	defer r.remoteClustersMutex.Unlock()
+
+	for _, remoteClusters := range r.remoteClusters {
+		remoteClusters.close()
+	}
+}
+
+func (r *LightHouseController) OnAdd(clusterID string, kubeConfig *rest.Config) {
+	klog.Infof("Adding cluster: %s", clusterID)
+	// startNewServiceWatcher acquires the lock before accessing the map.
+	r.startNewServiceWatcher(clusterID, kubeConfig)
+}
+
+func (r *LightHouseController) OnUpdate(clusterID string, kubeConfig *rest.Config) {
+	klog.Infof("Updating cluster: %s", clusterID)
+	r.remoteClustersMutex.Lock()
+	defer r.remoteClustersMutex.Unlock()
+	r.removeServiceWatcher(clusterID)
+	r.startNewServiceWatcher(clusterID, kubeConfig)
+}
+
+func (r *LightHouseController) OnRemove(clusterID string) {
+	r.remoteClustersMutex.Lock()
+	defer r.remoteClustersMutex.Unlock()
+	r.removeServiceWatcher(clusterID)
+}
+
+func (r *LightHouseController) startNewServiceWatcher(clusterID string, kubeConfig *rest.Config) {
+	clientSet, err := r.newClientset(kubeConfig)
+	if err != nil {
+		klog.Errorf("error creating client set for cluster %s: %s", clusterID, err.Error())
+		return
+	}
+	serviceInformer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options meta_v1.ListOptions) (runtime.Object, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options meta_v1.ListOptions) (watch.Interface, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Service{},
+		0,
+		cache.Indexers{},
+	)
+	remoteCluster := &remoteCluster{
+		stopCh:              make(chan struct{}),
+		clusterID:           clusterID,
+		serviceInformer:     serviceInformer,
+		queue:               workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		remoteServices:      r.remoteServices,
+		federator:           r.federator,
+		remoteServicesMutex: &sync.Mutex{},
+	}
+	r.remoteClustersMutex.Lock()
+	defer r.remoteClustersMutex.Unlock()
```

Comment:
This needs to be removed as the callers now `Lock/Unlock` -  note `Lock` is not re-entrant.